server/util/grpc_client/grpc_client.go

```diff
@@ -216,3 +224,55 @@ func CommonGRPCClientOptions() []grpc.DialOption {
 		}),
 	}
 }
+
+type grpcClientConnPoolCache struct {
+	env         environment.Env
+	connMutex   *sync.Mutex
+	connPoolMap map[string]interfaces.ClosableClientConn
+}
+
+func RegisterConnPoolCache(env environment.Env) {
+	c := NewGrpcClientConnPoolCache(env)
+	env.SetGrpcClientConnPoolCache(c)
+}
+
+func isPermittedForDial(target string) bool {
+	u, err := url.Parse(target)
+	if err != nil {
+		return false
+	}
+	if cache_api_url.String() == "" {
+		return true
+	}
+
+	return u.Hostname() == "localhost" || (urlutil.GetDomain(u.Hostname()) == urlutil.GetDomain(cache_api_url.WithPath("").Hostname()))
```

Comment:
this logic seems a bit too high level to live in a cache implementation? what do you think about moving this to the byte_stream_client code?